### PR TITLE
Move easy things from `transaction` to `primitives`

### PR DESCRIPTION
**Pull request by tcharding**
Move hash wrapper types and `transaction::Version`.

### CODE REVIEW

**Comment by tcharding** [comment]
Will fix tomorrow.

**Comment by tcharding** [comment]
> ~Nothing controversial here.~

I should be careful what I claim last thing in the afternoon.

File: primitives/src/transaction.rs [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+}
+
+impl Wtxid {
+    /// The "all zeros" wTXID.
+    ///
+    /// This is used as the wTXID for the coinbase transaction when constructing blocks,
+    /// since the coinbase transaction contains a commitment to all transactions' wTXIDs
+    /// but naturally cannot commit to its own. It is not a real wTXID and should not be
+    /// used in other contexts.
+    pub fn all_zeros() -> Self { Self::from_byte_array([0; 32]) }
```
Comment (Kixunil) [inline primitives/src/transaction.rs:51]: Makes me think if we want to rename these to something like `new_coinbase_input`.

Also they don't need to be moved. You can implement them in extension trait by calling `from_byte_array([0; 32])`

File: primitives/src/transaction.rs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+}
+
+/// Trait that abstracts over a transaction identifier i.e., `Txid` and `Wtxid`.
+pub trait TxIdentifier: sealed::Sealed + AsRef<[u8]> {}
```
Comment (Kixunil) [inline primitives/src/transaction.rs:55]: Why is this needed in `primitives`? It's only provided for one obscure function in `bitcoin`.

File: primitives/src/transaction.rs [hunk 3]
```diff
@@ -56,3 +58,34 @@ mod sealed {
...
+    pub const ONE: Self = Self(1);
+
+    /// The second Bitcoin transaction version (post-BIP-68).
+    pub const TWO: Self = Self(2);
+
+    /// Creates a non-standard transaction version.
+    pub fn non_standard(version: i32) -> Version { Self(version) }
+
+    /// Returns true if this transaction version number is considered standard.
+    pub fn is_standard(&self) -> bool { *self == Version::ONE || *self == Version::TWO }
```
Comment (Kixunil) [inline primitives/src/transaction.rs:90]: Way too much API for something that has a public field. But also `i32` bothers me. It's a stupid C++ism that we ought to support for compatibility but keeping the field as `i32` is ridiculous. What does negative version even mean? It should've been `u32` even in Core.

File: primitives/src/transaction.rs [hunk 4]
```diff
@@ -89,3 +93,133 @@ impl Version {
...
+
+    /// Creates a "null" `OutPoint`.
+    ///
+    /// This value is used for coinbase transactions because they don't have any previous outputs.
+    #[inline]
+    pub fn null() -> OutPoint { OutPoint { txid: Txid::all_zeros(), vout: u32::MAX } }
+
+    /// Checks if an `OutPoint` is "null".
+    #[inline]
+    pub fn is_null(&self) -> bool { *self == OutPoint::null() }
```
Comment (Kixunil) [inline primitives/src/transaction.rs:128]: Again, way too much API moved and also why is it called `null` here but `all_zeros` in other hashes?

File: primitives/src/transaction.rs [hunk 5]
```diff
@@ -89,3 +93,133 @@ impl Version {
...
+}
+
+impl fmt::Display for OutPoint {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}:{}", self.txid, self.vout)
+    }
+}
+
+#[cfg(feature = "alloc")]       // Because of `parse::ParseIntError`.
+impl core::str::FromStr for OutPoint {
```
Comment (Kixunil) [inline primitives/src/transaction.rs:142]: I object to this impl existing unless the format is super-wide standard. I've seen it used at multiple places but I'm not really 100% sure it's sensible.

File: primitives/src/transaction.rs [hunk 6]
```diff
@@ -89,3 +93,133 @@ impl Version {
...
+    /// Error in TXID part.
+    Txid(hex::HexToArrayError),
+    /// Error in vout part.
+    Vout(parse::ParseIntError),
+    /// Error in general format.
+    Format,
+    /// Size exceeds max.
+    TooLong,
+    /// Vout part is not strictly numeric without leading zeroes.
+    VoutNotCanonical,
```
Comment (Kixunil) [inline primitives/src/transaction.rs:193]: Hmm, this suggests that the encoding is actually standardized.

File: primitives/src/transaction.rs (hunk 4 above)
Comment (apoelstra) [reply primitives/src/transaction.rs:128]: An outpoint is not a hash. A "null outpoint", which is a standard term among people manually constructing transactions out of primitives, is one with an all-zeros hash and an all-ones vout.

File: primitives/src/transaction.rs (hunk 5 above)
Comment (apoelstra) [reply primitives/src/transaction.rs:142]: Assuming this is the `txid:vout` format, this is super common and used all over the place. There is also no other string format for outpoints that I have ever seen.

File: primitives/src/transaction.rs (hunk 6 above)
Comment (apoelstra) [reply primitives/src/transaction.rs:193]: It's not (I checked Core and it appears to not only accept leading zeros, but also `+` and `-`s lol). But this "canonical" requirement is there so that we can write fuzztests that check there is a 1-to-1 mapping between outpoints and their string representations.

I'd really rather not drop the canonicity requirement (which you may be the first person ever to even notice it *was* a requirement) because it'll eliminate that category of fuzztests.

File: primitives/src/transaction.rs (hunk 4 above)
Comment (Kixunil) [reply primitives/src/transaction.rs:128]: Ah, true, that makes sense.

File: primitives/src/transaction.rs (hunk 5 above)
Comment (Kixunil) [reply primitives/src/transaction.rs:142]: I vaguely remember seeing other things but I don't remember what exactly it was. Probably a different separator or no separator. Anyway, it's quite likely that this is fine.

File: primitives/src/transaction.rs (hunk 6 above)
Comment (Kixunil) [reply primitives/src/transaction.rs:193]: Can't we just filter out fuzztests differently?

File: primitives/src/transaction.rs (hunk 5 above)
Comment (apoelstra) [reply primitives/src/transaction.rs:142]: The `bitcoin-tx` tool uses this format to take outpoints on the command-line. The format <64 hex>:<number> is also widely recognized as being an outpoint, and appears on slides and on mailing list posts etc.

The format is also parsed by `hal`, the Liquid functionary code, and probably esplora ... though those don't "really" count since they are all using rust-bitcoin.

File: primitives/src/transaction.rs (hunk 6 above)
Comment (apoelstra) [reply primitives/src/transaction.rs:193]: Yes, in principle it's always possible to enumerate all forms of non-canonicity and filter for them (or re-canonicalize serializations, or define a weak notion of "equality").

But it's at least an order of magnitude more work than the "build an object, serialize and parse it, assert equality" type fuzztests that we typically have because they're so easy to write.

Comment (Kixunil) [reply primitives/src/transaction.rs:193]: What really bugs me is the inconsistency among implementations. In extreme cases such things can lead to vulnerabilities (though I have hard time imagining how they'd be abused here). OTOH restricting formats is usually a good thing.

**Comment by storopoli** [comment]
typo in PR title "transcation"...

**Comment by tcharding** [comment]
In keeping with the PR title I removed the last patch (the `OutPoint` stuff) since it has proven to be controversial, will do it separately using review comments.

**Comment by tcharding** [comment]
Ugh bother, `TxIdentifier` - my bad.

File: primitives/src/transaction.rs (hunk 1 above)
Comment (tcharding) [reply primitives/src/transaction.rs:47]: Resolved, this relates to outpoint.

File: primitives/src/transaction.rs (hunk 2 above)
Comment (tcharding) [reply primitives/src/transaction.rs]: Resolved.

File: primitives/src/transaction.rs (hunk 3 above)
Comment (tcharding) [reply primitives/src/transaction.rs]: I just noticed this is unresolved, now these are left behind in an extension trait.

**Comment by tcharding** [comment]
Can this go in like this or do you guys want me to split it into two PRs?

File: primitives/src/transaction.rs (hunk 1 above)
Comment (Kixunil) [reply primitives/src/transaction.rs:47]: `OutPoint` can just all `from_byte_array([0; 32])`. No need to move potentially-controversial methods.

**Review by Kixunil — COMMENTED** [review]
I'm convinced that the `null` method on `OutPoint` has a good name but that doesn't imply `all_zeros` is a good name on `Txid` and `Wtxid`. Should we call them `for_coinbase_txin` or something descriptive like that?

**Comment by apoelstra** [comment]
I don't think `null` is obvious or discoverable.

I'd suggest using `coinbase` for all of `OutPoint` and `Txid`/`Wtxid`. We should add doc aliases to `null` and `all_zeros`.

**Comment by tcharding** [comment]
> Move easy things from `transaction` to `primitives`

Was a hopeful title, wasn't it. I'm going to split this up into even more painfully small pieces.
